From fc4950160309a9a5c93559513344e75ab7beb027 Mon Sep 17 00:00:00 2001
From: jeebuscrossaint <thediamond270@gmail.com>
Date: Sat, 15 Aug 2026 21:30:47 -0400
Subject: [PATCH 14/29] patch: alwayscenter

---
 dwl.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dwl.c b/dwl.c
index 4fa5c90..27054ce 100644
--- a/dwl.c
+++ b/dwl.c
@@ -3035,6 +3035,8 @@ setmon(Client *c, Monitor *m, uint32_t newtags)
 		/* Make sure window actually overlaps with the monitor */
 		resize(c, c->geom, 0);
 		c->tags = newtags ? newtags : m->tagset[m->seltags]; /* assign tags of target monitor */
+		c->prev.x = (m->w.width - c->prev.width) / 2 + m->m.x;
+		c->prev.y = (m->w.height - c->prev.height) / 2 + m->m.y;
 		setfullscreen(c, c->isfullscreen); /* This will call arrange(c->mon) */
 		setfloating(c, c->isfloating);
 	}
-- 
2.55.0

