From b5a7e576989661e62c6cc01d7191ba3cb36c5d4b Mon Sep 17 00:00:00 2001
From: jeebuscrossaint <thediamond270@gmail.com>
Date: Sat, 15 Aug 2026 21:37:37 -0400
Subject: [PATCH 22/29] patch: fullscreenadaptivesync - auto VRR when a client
 is fullscreen

---
 config.def.h |  1 +
 dwl.c        | 53 ++++++++++++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 54 insertions(+)

diff --git a/config.def.h b/config.def.h
index 6306cdb..9f298cc 100644
--- a/config.def.h
+++ b/config.def.h
@@ -218,6 +218,7 @@ static const Key keys[] = {
 	{ MODKEY,                    XKB_KEY_x,           toggleswallow,    {0} },
 	{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_A,           toggleautoswallow,{0} },
 	{ MODKEY,                    XKB_KEY_v,           togglepointer,    {0} },
+	{ MODKEY,                    XKB_KEY_F5,          togglefullscreenadaptivesync, {0} },
 	{ MODKEY,                    XKB_KEY_0,           view,             {.ui = ~0} },
 	{ MODKEY|WLR_MODIFIER_SHIFT, XKB_KEY_parenright,  tag,              {.ui = ~0} },
 	{ MODKEY,                    XKB_KEY_comma,       focusmon,         {.i = WLR_DIRECTION_LEFT} },
diff --git a/dwl.c b/dwl.c
index eda1594..0bada49 100644
--- a/dwl.c
+++ b/dwl.c
@@ -395,6 +395,7 @@ static void requestmonstate(struct wl_listener *listener, void *data);
 static void resize(Client *c, struct wlr_box geo, int interact);
 static void run(char *startup_cmd);
 static void setcfact(const Arg *arg);
+static void set_adaptive_sync(Monitor *m, int enabled);
 static void setcursor(struct wl_listener *listener, void *data);
 static void setcursorshape(struct wl_listener *listener, void *data);
 static void setfloating(Client *c, int floating);
@@ -427,6 +428,7 @@ static void togglescratchpad(const Arg *arg);
 static void toggleswallow(const Arg *arg);
 static void toggleautoswallow(const Arg *arg);
 static void togglepointer(const Arg *arg);
+static void togglefullscreenadaptivesync(const Arg *arg);
 static void toggletag(const Arg *arg);
 static void toggleview(const Arg *arg);
 static void unlocksession(struct wl_listener *listener, void *data);
@@ -548,6 +550,8 @@ static struct wl_listener request_start_drag = {.notify = requeststartdrag};
 static struct wl_listener start_drag = {.notify = startdrag};
 static struct wl_listener new_session_lock = {.notify = locksession};
 
+static int fullscreen_adaptive_sync_enabled = 1;
+
 #ifdef XWAYLAND
 static void activatex11(struct wl_listener *listener, void *data);
 static void associatex11(struct wl_listener *listener, void *data);
@@ -3103,6 +3107,43 @@ setcfact(const Arg *arg)
 	arrange(selmon);
 }
 
+void
+set_adaptive_sync(Monitor *m, int enable)
+{
+	struct wlr_output_state state;
+	struct wlr_output_configuration_v1 *config;
+	struct wlr_output_configuration_head_v1 *config_head;
+
+	if (!m || !m->wlr_output || !m->wlr_output->enabled
+			|| !fullscreen_adaptive_sync_enabled)
+		return;
+
+	if (enable && m->wlr_output->adaptive_sync_status
+			== WLR_OUTPUT_ADAPTIVE_SYNC_ENABLED)
+		return;
+
+	if (!enable && m->wlr_output->adaptive_sync_status
+			== WLR_OUTPUT_ADAPTIVE_SYNC_DISABLED)
+		return;
+
+	config = wlr_output_configuration_v1_create();
+	config_head = wlr_output_configuration_head_v1_create(config, m->wlr_output);
+
+	/* Set and commit the adaptive sync state change */
+	wlr_output_state_init(&state);
+	wlr_output_state_set_adaptive_sync_enabled(&state, enable);
+	if (!wlr_output_commit_state(m->wlr_output, &state)) {
+		wlr_output_state_finish(&state);
+		wlr_output_configuration_v1_destroy(config);
+		return;
+	}
+	wlr_output_state_finish(&state);
+
+	/* Broadcast the adaptive sync state change to output_mgr */
+	config_head->state.adaptive_sync_enabled = enable;
+	wlr_output_manager_v1_set_configuration(output_mgr, config);
+}
+
 void
 setcursor(struct wl_listener *listener, void *data)
 {
@@ -3191,10 +3232,12 @@ setfullscreen(Client *c, int fullscreen)
 	if (fullscreen) {
 		c->prev = c->geom;
 		resize(c, c->mon->m, 0);
+		set_adaptive_sync(c->mon, 1);
 	} else {
 		/* restore previous size instead of arrange for floating windows since
 		 * client positions are set by the user and cannot be recalculated */
 		resize(c, c->prev, 0);
+		set_adaptive_sync(c->mon, 0);
 	}
 	arrange(c->mon);
 	drawbars();
@@ -3949,6 +3992,12 @@ togglepointer(const Arg *arg)
 	);
 }
 
+void
+togglefullscreenadaptivesync(const Arg *arg)
+{
+	fullscreen_adaptive_sync_enabled = !fullscreen_adaptive_sync_enabled;
+}
+
 void
 toggletag(const Arg *arg)
 {
@@ -4004,6 +4053,7 @@ unmapnotify(struct wl_listener *listener, void *data)
 {
 	/* Called when the surface is unmapped, and should no longer be shown. */
 	Client *c = wl_container_of(listener, c, unmap);
+	Monitor *lastmon = c->mon; // fullscreen_adaptive_sync
 	if (c == grabc) {
 		cursor_mode = CurNormal;
 		grabc = NULL;
@@ -4033,6 +4083,9 @@ unmapnotify(struct wl_listener *listener, void *data)
 		setmon(c, NULL, 0);
 		wl_list_remove(&c->flink);
 	}
+	/* Toggle adaptive sync off when fullscreen client is unmapped */
+	if (c->isfullscreen)
+		set_adaptive_sync(lastmon, 0);
 
 	wlr_scene_node_destroy(&c->scene->node);
 	drawbars();
-- 
2.55.0

